### DISCUSSION TITLE
Handling failures in Init

### DISCUSSION THREAD

**Discussion by ZeroDeltaAlpha** [charmbracelet/bubbletea#623] [Q&A]
Hi I'm writing an app that does data fetching in the init func using http. 

The data fetching func returns a Struct and a error type, in the case of a failure I send back an empty Struct and an err.

 i run this using tea.Sequence, what is the best way to handle if err is not nil on init and take appropriate action ?

**Comment by bashbunni** [comment]
You can have a `tea.Msg` that is a struct holding your outputs from the data fetch. Then return that message. A `tea.Msg` is any data structure, you just need to specify in the function signature of your function as the return value, then you can use your function as a `tea.Cmd`

```go
type HttpResponse struct {
    data []byte // change this to your struct if you want, or make your struct include the error
    err error
}

func FetchData(url string) tea.Msg {
    resp, err := http.Get("http://example.com/")
    if err != nil {
	    return HttpResponse{[]byte{}, err}
    }
    defer resp.Body.Close()
    body, err := io.ReadAll(resp.Body)
    return HttpResponse{body, err}
}
```

I hope that helps!

**Reply by ZeroDeltaAlpha** [reply]
Hi @bashbunni my issue is more like the following
```go
func (m model) Init() tea.Cmd {
	cmds := make([]tea.Cmd, 0)
	cmds = append(cmds, List(fetchData(m.cfg)), tea.EnterAltScreen)
	return tea.Sequence(cmds...)
}
```

fetchData returns a list of items that could be empty and err. Is there a way i can surface this err from the Init function ?

Thanks

**Reply by bashbunni** [reply]
if you're wanting it to run asynchronously, you'll have to call those `tea.Cmds` like you're doing, then handle the errors in your `Update` when you receive the response from the function. You can return the response as a struct, then check for that kind of `tea.Msg` in your `Update`. Outside of that, you'd have to handle the error before you start your Bubble Tea program, but how I just described is how you handle errors within a Bubble Tea program.

Here's an example with `ErrMsg`: https://github.com/charmbracelet/bubbletea/blob/6b77c8fc10d43195ab29e6e09f93272623ce4e9c/examples/http/main.go#L52

I hope that helps!

**Reply by ZeroDeltaAlpha** [reply]
Awesome thanks @bashbunni I think i got that part nailed down, thank you for the help 😄
